### ISSUE TITLE
Authentication Enhancement

### ISSUE THREAD

**Issue by NilsG-S** [TTUSDC/CPCEEDWebApp#20]
Features:
- Input checking
- Highlighting for incorrect fields

**Comment by asclines** [comment]
Perhaps we could even do some stuff client-side? Like checking if the email is a valid email before even sending a request to the backend?

**Comment by NilsG-S** [comment]
That's sort of what I was thinking. It might actually be easier than parsing the error messages.

**Comment by asclines** [comment]
Maybe, the errors received from Firebase are designed to be parsed so it might not be too bad.

**Comment by NilsG-S** [comment]
We should discuss what input values will be considered valid for email and password.

**Comment by asclines** [comment]
@NilsG-S what do you mean?

**Comment by NilsG-S** [comment]
@asclines 
I haven't found any clean ways of checking whether [emails](http://stackoverflow.com/questions/46155/validate-email-address-in-javascript) are valid, so I'm not sure what criteria we will be using. Should we just check that it ends in @ttu.edu?

We're no longer using Firebase, so our password standards aren't already determined for us. We'll need to decide what passwords are acceptable.

Student IDs should be easier, aren't they all the same length?

**Comment by asclines** [comment]
@NilsG-S 
For the emails we could just have an automated email verification? Or some parser that checks for valid email.

Student IDs will be easier unless we want to check that it is a legit Student ID but that might be better left to the admins to check. 

As for passwords, good point does @the-pat have an opinion?

**Comment by NilsG-S** [comment]
@asclines Yeah, the link I put in my comment has instructions for how to use regular expressions for email verification. I'm just not sure what emails we will be accepting. If we just use TTU email addresses it'll be relatively easy to make a regex that verifies them.

**Comment by NilsG-S** [comment]
Okay, so @mariots used his contacts to get some information on how TTU email addresses are structured. For the most part it's as easy as I thought. Basically they're dot separated names, which can include hyphens. If there is a name collision the middle initial is added. If there is a middle initial collision, the whole middle name is added. If there is a collision after that, we're not really sure what happens.

Luckily the ending will always be @ttu.edu for students and administrators.

The more serious issue is that there is a way for people to make a custom email address. If we want to handle all cases, we might have to use a general regex.

**Comment by asclines** [comment]
would we really need much more validation than checking `@ttu.edu`?

**Comment by NilsG-S** [comment]
That would work for me

**Comment by the-pat** [comment]
@asclines and @NilsG-S 

Some general password policies are:

- minimum length of 8 characters
- allow all valid ASCII characters

How do y'all feel about enforcing a password with _at least one_ upper-case, lower-case, number, and special character?

Also, I don't like the idea of a maximum password length, but the maximum BSON size is [16 megabytes](https://docs.mongodb.com/manual/reference/limits/). If we want to impose a maximum password length, I'd want to lower bound to be at least 64 characters.

Thoughts?

**Comment by asclines** [comment]
I have no real preference on this so I'll defer my opinion to @NilsG-S

**Comment by NilsG-S** [comment]
@the-pat What do you mean by special characters? Anything that isn't a word character (A-Z, a-z, 0-9, and underscore)?
